components/support/base/src/main/java/mozilla/components/support/base/concurrent/RunWhenReadyQueue.kt

```diff
@@ -0,0 +1,62 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package mozilla.components.support.base.concurrent
+
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.MainScope
+import kotlinx.coroutines.launch
+import java.util.concurrent.CopyOnWriteArrayList
+import java.util.concurrent.atomic.AtomicBoolean
+
+/**
+ * A queue that acts as a gate, either executing tasks right away if the queue is marked as "ready",
+ * i.e. gate is open, or queues them to be executed whenever the queue is marked as ready in the
+ * future, i.e. gate becomes open.
+ */
+class RunWhenReadyQueue(
+    private val scope: CoroutineScope = MainScope()
+) {
+
+    private val tasks = CopyOnWriteArrayList<() -> Unit>()
+    private val isReady = AtomicBoolean(false)
+
+    /**
+     * Was this queue ever marked as 'ready' via a call to [ready]?
+     *
+     * @return Boolean value indicating if this queue is 'ready'.
+     */
+    fun isReady(): Boolean = isReady.get()
+
+    /**
+     * Runs the [task] if this queue is marked as ready, or queues it for later execution.
+     *
+     * @param task: The task to run now if queue is ready or queue for later execution.
+     */
+    fun runIfReadyOrQueue(task: () -> Unit) {
+        if (isReady.get()) {
+            scope.launch { task.invoke() }
+        } else {
+            tasks.add(task)
+        }
```

Comment:
This part doesn't seem thread-safe is `runIfReadyOrQueue()` and `ready()` are called on different threads? I think we would need to synchronize here? Or alternatively assert that we are on the main thread if that is what we expect.

Reply:
Would `AtomicBoolean` handle the threads? cc @grigoryk who worked on the class

Reply:
I think what _could_ happen here is that `tasks` is modified by `runIfReadyOrQueue` while we're iterating through it in `ready`. This won't result in an exception thanks to `CopyOnWriteArrayList`, but we will drop that task on the floor in the `tasks.clear` call right after the iteration, I think.

A `synchronized` block wrapping the iteration/clearing of `tasks`, and wrapping `tasks.add` should be fine, though. With that in place, you can switch to a regular array, as well.